### PR TITLE
Add function to setup attrs of an xml object

### PR DESCRIPTION
**Pull request by chloerh**
The function allows us to setup attributes of an xml object with
a dict-like param.

Signed-off-by: Haijiao Zhao <haizhao@redhat.com>

### CODE REVIEW

**Comment by chloerh** [comment]
Test with vmxml
```
>>> from virttest.libvirt_xml import vm_xml
>>> vmxml = vm_xml.VMXML.new_from_dumpxml('avocado-vt-vm1')
>>> print(vmxml)
<domain type="kvm">
  <name>avocado-vt-vm1</name>
...
  <memory unit="KiB">1048576</memory>
  <currentMemory unit="KiB">1048576</currentMemory>
  <vcpu placement="static">2</vcpu>
  <os>
...
</domain>

>>> vm_attrs = {'max_mem_rt': 4096, 'max_mem_rt_slots': 16, 'max_mem_rt_unit': 'M', 'vcpu': 4, 'cpu': {'numa_cell': [{'id': '0', 'cpus': '0-1', 'memory': '512', 'unit': 'M'}, {'id': '1', 'cpus': '2-3', 'memory': '512', 'unit': 'M'}]}}
>>> vmxml.setup_attrs(**vm_attrs)
>>> print(vmxml)
<domain type="kvm">
  <name>avocado-vt-vm1</name>
...
<maxMemory slots="16" unit="M">4096</maxMemory><cpu check="partial" mode="host-model">
    <feature name="vmx" policy="disable" />
  <numa><cell cpus="0-1" id="0" memory="512" unit="M" /><cell cpus="2-3" id="1" memory="512" unit="M" /></numa></cpu></domain>
```

Test with disk xml:
```
>>> from virttest.libvirt_xml.devices.disk import Disk
>>> d1 = Disk('network')
>>> d1_attrs = {'device': 'disk', 'driver': {'name': 'qemu', 'type': 'raw'}, 'source': {'attrs': {'protocol': 'iscsi', 'name': 'iqn.2013-07.com.example:iscsi-nopool/2'}, 'hosts': [{'name': 'example.com', 'port': '3260'}], 'auth': {'auth_user': 'myuser', 'secret_type': 'iscsi', 'secret_usage': 'libvirtiscsi'}}, 'target': {'dev': 'vda', 'bus': 'virtio'}}
>>> d1.setup_attrs(**d1_attrs)
>>> print(d1)
<?xml version='1.0' encoding='UTF-8'?>
<disk device="disk" type="network">
<driver name="qemu" type="raw" />
<source name="iqn.2013-07.com.example:iscsi-nopool/2" protocol="iscsi">
<host name="example.com" port="3260" />
<auth username="myuser">
<secret type="iscsi" usage="libvirtiscsi" />
</auth>
</source>
<target bus="virtio" dev="vda" />
</disk>
>>> 

```

**Comment by chloerh** [comment]
@kylazhang @dzhengfy @Yingshun Could you help review this? Thank you.

File: virttest/libvirt_xml/base.py [hunk 1]
```diff
@@ -233,6 +233,57 @@ def virt_xml_validate(filename, schema_name=None):
...
+            <memory model="dimm">
+                <target>
+                    <size unit="KiB">524288</size><node>0</node>
+                </target>
+                <alias name="dimm2" />
+                <address base="0x120000000" slot="2" type="dimm" />
+            </memory>
+
+        """
+        reset_all = True if attrs.get('reset_all') is True else False
```
Comment (kylazhang) [inline virttest/libvirt_xml/base.py:264]: I think it is enough to set like this
```
reset_all = attrs.get('reset_all') is True
```

File: virttest/libvirt_xml/base.py [hunk 2]
```diff
@@ -233,6 +233,57 @@ def virt_xml_validate(filename, schema_name=None):
...
+
+            # Is XMLElementNest or not
+            subclass = get_func.get('subclass')
+            if subclass is None:
+                setattr(self, key, value)
+            else:
+                # Element tag not found or reset_all
+                if reset_all or not self.xmltreefile.find(key):
+                    subclass_dargs = get_func.get('subclass_dargs')
+                    target_obj = subclass(**subclass_dargs)
```
Comment (kylazhang) [inline virttest/libvirt_xml/base.py:281]: Sorry can not understand here, could you please explain more here with comments? Thanks

File: virttest/libvirt_xml/base.py (hunk 1 above)
Comment (Yingshun) [inline virttest/libvirt_xml/base.py:264]: Would you mind providing an example of reset_all usages? 
I faced a problem as below if set it as True in attrs. Am I wrong?
```
>>> dargs={"managed":"yes", "mode":"subsystem", "type":"pci",'reset_all': True,"source":{'untyped_address':{"bus":"0x3b","domain":"0x0000","function":"0x1", "slot":"0x00"}}}
>>> dargs.get('reset_all')
True
>>> hostdev2.setup_attrs(**dargs)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/var/tmp/libvirt-ci/runtest/avocado-vt/avocado-vt/virttest/libvirt_xml/base.py", line 265, in setup_attrs
    raise AttributeError('Cannot set attribute "%s" to %s object.'
AttributeError: Cannot set attribute "reset_all" to <class 'virttest.libvirt_xml.devices.hostdev.Hostdev'> object.There is no such attribute.

```

Comment (chloerh) [reply virttest/libvirt_xml/base.py:264]: You're right, 'reset_all' should be removed from attrs after we have its value. Thanks for finding the bug :+1:

Comment (chloerh) [reply virttest/libvirt_xml/base.py:264]: Yes, you're right

File: virttest/libvirt_xml/base.py (hunk 2 above)
Comment (chloerh) [reply virttest/libvirt_xml/base.py:283]: Ok

File: virttest/libvirt_xml/base.py (hunk 1 above)
Comment (Yingshun) [reply virttest/libvirt_xml/base.py]: It works now. LGTM.
```
>>> hostdev1=hostdev.Hostdev()
>>> hostdev1.setup_attrs(**dargs)
>>> print(hostdev1)
<?xml version='1.0' encoding='UTF-8'?>
<hostdev managed="yes" mode="subsystem" type="pci"><source><address bus="0x3b" domain="0x0000" function="0x1" slot="0x00" /></source></hostdev>
>>> 
```

**Review by Yingshun — APPROVED** [review]

File: virttest/libvirt_xml/base.py (hunk 1 above)
Comment (chloerh) [reply virttest/libvirt_xml/base.py]: Thanks again for helping test with a different xml type!

**Review by kylazhang — COMMENTED** [review]
Please update the comments as we discussed for reset_all and explanation for the if branch, thanks

**Comment by chloerh** [comment]
@kylazhang comments updated.

**Review by kylazhang — APPROVED** [review]

File: virttest/libvirt_xml/base.py [hunk 3]
```diff
@@ -233,6 +233,66 @@ def virt_xml_validate(filename, schema_name=None):
...
+                    <size unit="KiB">524288</size><node>0</node>
+                </target>
+                <alias name="dimm2" />
+                <address base="0x120000000" slot="2" type="dimm" />
+            </memory>
+
+        """
+        # Whether to keep the sub-xml instance and modify it
+        # or completely re-create one
+        reset_all = attrs.get('reset_all') is True
```
Comment (dzhengfy) [inline virttest/libvirt_xml/base.py:266]: Please add the docstring for 'reset_all' usage.

**Review by dzhengfy — COMMENTED** [review]
Others LGTM

**Review by dzhengfy — APPROVED** [review]
LGTM

**Pull request merged**